packages/block-editor/src/components/link-control/use-remote-url-data.js

```diff
@@ -0,0 +1,88 @@
+/**
+ * Internal dependencies
+ */
+import { store as blockEditorStore } from '../../store';
+import makeCancelable from './make-cancelable';
+
+/**
+ * WordPress dependencies
+ */
+import { useSelect } from '@wordpress/data';
+import { useState, useEffect, useRef, useCallback } from '@wordpress/element';
+
+function useRemoteUrlData( url ) {
+	const cancelableFetch = useRef();
+
+	const [ richData, setRichData ] = useState( null );
+	const [ isFetching, setIsFetching ] = useState( false );
+
+	const { fetchRemoteUrlData } = useSelect( ( select ) => {
+		const { getSettings } = select( blockEditorStore );
+		return {
+			fetchRemoteUrlData: getSettings().__experimentalFetchRemoteUrlData,
+		};
+	}, [] );
+
+	const cancelPendingFetch = useCallback( () => {
+		if ( cancelableFetch.current ) {
+			cancelableFetch.current.cancel();
+		}
+	}, [ cancelableFetch ] );
+
+	/**
+	 * Cancel any pending requests that were made for
+	 * stale URL values. If the component does not unmount
+	 * we must handle cancelling the current request if
+	 * the URL changes otherwise we will see stale data.
+	 */
+	useEffect( () => {
+		cancelPendingFetch();
+	}, [ url ] );
+
+	/**
+	 * Cancel any pending requests on "unmount"
+	 */
+	useEffect( () => {
+		return () => {
+			cancelPendingFetch();
+		};
+	}, [] );
+
+	useEffect( () => {
+		const fetchRichData = async () => {
+			setIsFetching( true );
+
+			// Clear the data if the URL changes to avoid stale data in hook consumer.
+			setRichData( null );
```

Comment:
I believe this will cause two different renders since they are in a callback which React doesn't control. A better idea would be to leverage `useReducer` instead and encapsulate them together. It's also more extensible and we can add error handling too in the future.